### PR TITLE
[9.x] Alternative Mailable Syntax

### PR DESCRIPTION
**Pull request by taylorotwell**
Currently, mailables are "configured" via a `build` method. Within this method, various methods on the parent mailable class are invoked to specify the structure of the mailable:

```php
public function build()
{
    return $this->from('example@example.com', 'Example')
                ->subject('Order Shipped')
                ->view('emails.orders.shipped');
}
```

While this works fine, over the last couple of years it has always rubbed me the wrong way. It feels a lot like an initialization method or a two-step constructor. The mailable object can't be inspected in any useful way until that method is invoked. There aren't really any other types of objects in Laravel user-land that work this way or follow this pattern.

This PR offers an alternative syntax for defining mailables in which several discrete methods may be defined that each return configuration style objects that describe the structure of the mailable:

```php
<?php

namespace App\Mail;

use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Mail\Mailable;
use Illuminate\Mail\Mailables\Address;
use Illuminate\Mail\Mailables\Attachment;
use Illuminate\Mail\Mailables\Content;
use Illuminate\Mail\Mailables\Envelope;
use Illuminate\Queue\SerializesModels;

class InvoicePaid extends Mailable
{
    use Queueable, SerializesModels;

    /**
     * Create a new message instance.
     *
     * @return void
     */
    public function __construct()
    {
        //
    }

    /**
     * Get the message envelope.
     *
     * @return \Illuminate\Mail\Mailables\Envelope
     */
    public function envelope()
    {
        return new Envelope(
            subject: 'Invoice Paid',
            cc: [new Address('foo@example.com', 'Example Name')],
            tags: [],
            metadata: [],
        );
    }

    /**
     * Get the message content definition.
     *
     * @return \Illuminate\Mail\Mailables\Content
     */
    public function content()
    {
        return new Content(
            view: 'html-view-name',
            text: 'text-view-name',
        );
    }

    /**
     * Get the attachments for the message.
     *
     * @return \Illuminate\Mail\Mailables\Attachment[]
     */
    public function attachments()
    {
        return [
            Attachment::fromPath('/path/to/file'),
        ];
    }
}
```

**Note:** Only the `content` method is really required when defining mailables in this fashion.

A more minimal example would look like the following:

```php
<?php

namespace App\Mail;

use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Mail\Mailable;
use Illuminate\Mail\Mailables\Content;
use Illuminate\Queue\SerializesModels;

class InvoicePaid extends Mailable
{
    use Queueable, SerializesModels;

    /**
     * Get the message content definition.
     *
     * @return \Illuminate\Mail\Mailables\Content
     */
    public function content()
    {
        return new Content(
            markdown: 'invoice-paid',
        );
    }
}
```

The DX of this feature relies heavily on PHP 8 named arguments to provide a nice, clean experience when defining the configuration of the mailable object.

While the resulting mailable does end up having more lines of code overall, it feels a bit more like a real object and not a weird, two-step constructed thing. The `envelope`, `content`, and `attachments` methods may all be invoked as many times as needed and their results inspected like normal objects.

A few new methods have also been added to inspect mailable instances:

```php
$mailable->hasSubject('Foo');
$mailable->hasTag('foo');
$mailable->hasMetadata('key', 'value');
```

In addition, an optional `headers` method may be defined to add additional headers, including arbitrary text headers to the message:

```php
use Illuminate\Mail\Mailables\Headers;

public function headers()
{
    return new Headers(
        messageId: 'some-message-id@laravel.com',
        references: ['reference@laravel.com', 'another-reference@laravel.com'],
        text: [
            'X-Some-Header' => 'Some-Value',
        ],
    );
}
```

**The traditional way of defining mailables using the `build` method will not be removed.**

### CODE REVIEW

**Comment by Sammyjo20** [comment]
In fact methods will be really helpful to conditionally apply logic like, if in production always CC the sales email address.

**Comment by adityafullstackdeveloper** [comment]
Syntax looks great

**Comment by hailwood** [comment]
> In fact methods will be really helpful to conditionally apply logic like, if in production always CC the sales email address.

So perhaps as well as the methods we could add the `Conditionable` trait as well so we can use `when`/`unless` chaining?

**Comment by taylorotwell** [comment]
```php
cc: ['taylor@example.com'] + ($production ? ['second@example.com'] : []),
```

File: src/Illuminate/Mail/Mailables/Headers.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    public $text;
+
+    /**
+     * Create a new instance of headers for a message.
+     *
+     * @param  string|null  $messageId
+     * @param  array  $references
+     * @param  array  $text
+     * @return void
+     */
```
Comment (timacdonald) [inline src/Illuminate/Mail/Mailables/Headers.php:37]:
```suggestion
     * @return void
     *
     * @named-arguments-supported
     *
     */
```

**Comment by hailwood** [comment]
> ```
> cc: ['taylor@example.com'] + ($production ? ['second@example.com'] : []),
> ```

Sure that's possible but it's not exactly nice to scan, especially if you're using the full address syntax?
```php
return new Envelope(
    subject: 'Invoice Paid',
    cc: [new Address('taylor@example.com', 'Taylor Otwell')] + ($production ? [new Address('second@example.com', 'Second Address')] : []),
    tags: [],
    metadata: [],
);
```
I'd be more inclined to go
```php
$ccRecipients = [new Address('taylor@example.com', 'Taylor Otwell')];

if($production) {
    $ccRecipients[] = new Address('second@example.com', 'Second Address');
}

return new Envelope(
    subject: 'Invoice Paid',
    cc: $ccRecipients,
    tags: [],
    metadata: [],
);
```
But isn't it easier to understand if it's just
```php
$envelope = new Envelope(
    subject: 'Invoice Paid',
    cc: [new Address('taylor@example.com', 'Taylor Otwell')],
    tags: [],
    metadata: [],
));

if($production) {
    $envelope->addRecipient(new Address('second@example.com', 'Second Address'), type: 'cc');
}

return $envelope;
```
Or if conditionable then
```php
return (new Envelope(
    subject: 'Invoice Paid',
    cc: [new Address('taylor@example.com', 'Taylor Otwell')],
    tags: [],
    metadata: [],
)))->when($production, function(Envelope $envelope) {
    $envelope->addRecipient(new Address('second@example.com', 'Second Address'), type: 'cc');
});

File: src/Illuminate/Mail/Mailables/Envelope.php [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+     * @param  \Illuminate\Mail\Mailables\Address|string|null  $from
+     * @param  array  $to
+     * @param  array  $cc
+     * @param  array  $bcc
+     * @param  array  $replyTo
+     * @param  string|null  $subject
+     * @param  array  $tags
+     * @param  array  $metadata
+     * @return void
+     */
```
Comment (timacdonald) [inline src/Illuminate/Mail/Mailables/Envelope.php:75]:
```suggestion
     * @return void
     *
     * @named-arguments-supported
     *
     */
```

File: src/Illuminate/Mail/Mailables/Content.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * Create a new content definition.
+     *
+     * @param  string|null  $view
+     * @param  string|null  $html
+     * @param  string|null  $text
+     * @param  string|null  $markdown
+     * @param  array  $with
+     */
```
Comment (timacdonald) [inline src/Illuminate/Mail/Mailables/Content.php:52]:
```suggestion
     * @param  array  $with
     *
     * @named-arguments-supported
     *
     */
```

**Review by timacdonald — COMMENTED** [review]
Suggesting we explicitly mark constructors with named argument support.

This is a custom docblock, similar to what is seen in PHPUnit (however they are opting out of named argument support).

As Laravel does not support named arguments globally, this will serve as a marker for developers and maintainers that the feature is supported here.

File: src/Illuminate/Mail/Mailables/Envelope.php [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+     * @return array
+     */
+    protected function normalizeAddresses($addresses)
+    {
+        return collect($addresses)->map(function ($address) {
+            return is_string($address) ? new Address($address) : $address;
+        })->all();
+    }
+
+    /**
```
Comment (lioneaglesolutions) [inline src/Illuminate/Mail/Mailables/Envelope.php:101]: This isn't perfect but quickly threw it together... this is inline with the current mailable fluent methods. I'm not familiar with tags or meta-data but I'm sure it would be just as simple to add those 2 methods.

Suggestion to test also added and can confirm passing locally.

Wasn't sure if I could somehow commit this or do another PR so just added as a suggestion...

```suggestion
    /**
     * Add additional recipients to the message
     *
     * @param string|\Illuminate\Mail\Mailables\Address|array $address
     * @return $this
     */
    public function addTo($address)
    {
        $this->addRecipient($address, 'to');

        return $this;
    }

    /**
     * Add "cc" recipients to the message.
     *
     * @param string|\Illuminate\Mail\Mailables\Address|array $address
     * @return $this
     */
    public function addCc($address)
    {
        $this->addRecipient($address, 'cc');

        return $this;
    }

    /**
     * Add "bcc" recipients to the message.
     *
     * @param string|\Illuminate\Mail\Mailables\Address|array $address
     * @return $this
     */
    public function addBcc($address)
    {
        $this->addRecipient($address, 'bcc');

        return $this;
    }

    /**
     * Add "reply to" recipients to the message.
     *
     * @param string|\Illuminate\Mail\Mailables\Address|array $address
     * @return $this
     */
    public function addReplyTo($address)
    {
        $this->addRecipient($address, 'replyTo');

        return $this;
    }

    /**
     * Add additional recipients to the message
     *
     * @param string|\Illuminate\Mail\Mailables\Address|array $address
     * @param $type
     * @return $this
     * @throws \InvalidArgumentException
     */
    protected function addRecipient($address, $type) {
        if (!in_array($type, ['to', 'cc', 'bcc', 'replyTo'])) {
            throw new \InvalidArgumentException("$type is not a valid recipient type.");
        }

        $this->{$type} = [
            ...$this->{$type},
            ...$this->normalizeAddresses(is_array($address) ? $address : [$address])
        ];

        return $this;
    }
```

File: tests/Mail/MailableAlternativeSyntaxTest.php [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        $method->invoke($mailable);
+
+        $this->assertEquals('test-view', $mailable->view);
+        $this->assertEquals(['test-data-key' => 'test-data-value'], $mailable->viewData);
+        $this->assertEquals(2, count($mailable->to));
+        $this->assertEquals(1, count($mailable->cc));
+        $this->assertEquals(1, count($mailable->bcc));
+    }
+}
+
```
Comment (lioneaglesolutions) [inline tests/Mail/MailableAlternativeSyntaxTest.php:44]: Tests for suggestion Envelope fluent methods.

```suggestion
    public function testFluentEnvelopeMethods()
    {
        $mailable = new MailableWithFluentEnvelopeMethods;

        $this->assertTrue($mailable->hasTo('taylor@laravel.com'));
        $this->assertTrue($mailable->hasCc('adam@laravel.com'));
        $this->assertTrue($mailable->hasBcc('tyler@laravel.com'));
        $this->assertTrue($mailable->hasReplyTo('reply-to@laravel.com'));

        $this->assertTrue($mailable->hasTo('string@example.com'));
        $this->assertTrue($mailable->hasTo('array@example.com'));
        $this->assertTrue($mailable->hasTo('address@example.com'));
        $this->assertTrue($mailable->hasTo('address-array@example.com'));
        $this->assertTrue($mailable->hasCc('cc@example.com'));
        $this->assertTrue($mailable->hasBcc('bcc@example.com'));
        $this->assertTrue($mailable->hasReplyTo('reply-to@example.com'));
    }
}

class MailableWithFluentEnvelopeMethods extends Mailable
{
    public function envelope()
    {
        return (new Envelope(
            to: [new Address('taylor@laravel.com', 'Taylor Otwell')],
            cc: [new Address('adam@laravel.com', 'Adam Wathan')],
            bcc: [new Address('tyler@laravel.com', 'Tyler Blair')],
            replyTo: [new Address('reply-to@laravel.com', 'Reply To')],
            subject: 'Test Subject',
            tags: ['tag-1', 'tag-2'],
            metadata: ['test-meta' => 'test-meta-value'],
        ))
            ->addTo('string@example.com')
            ->addTo(['array@example.com'])
            ->addTo(new Address('address@example.com'))
            ->addTo([new Address('address-array@example.com')])
            ->addCc('cc@example.com')
            ->addBcc('bcc@example.com')
            ->addReplyTo('reply-to@example.com');
    }

    public function content()
    {
        return new Content(
            view: 'test-view',
            with: ['test-data-key' => 'test-data-value'],
        );
    }
}
```

**Review by lioneaglesolutions — COMMENTED** [review]
It would indeed be nice to add the fluent methods - have suggested some changes including the tests.

See;
- https://github.com/laravel/framework/pull/44462#issuecomment-1270280338
- https://github.com/laravel/framework/pull/44462#issuecomment-1270760347

**Comment by andrey-helldar** [comment]
For me personally, explicitly passing constructor properties is redundant and ugly.

So agree with @Sammyjo20 for [adding methods](https://github.com/laravel/framework/pull/44462#issuecomment-1270280338).

**Comment by nathanjansen** [comment]
Just another possible syntax.

```php
public function envelope()
{
    return new Envelope(
        subject: 'Invoice Paid',
        cc: Envelope\CC::make()
            ->add(new Address('foo@example.com', 'Example Name'))
            ->when($production, new Address('admin@example.com', 'Example Admin')),
        tags: [],
        metadata: [],
    );
}
```

**Comment by garygreen** [comment]
This is just "meh" to me. This PR is trying to get around the "two-step constructor" but it introduces [many more constructs in the process](https://github.com/laravel/framework/pull/44462/files#diff-76badd2790802d269067aad1c57c3eb75a83f0f8e3bddbb53db5832c818a4603R1378-R1381) in order to fully hydrate mail.

The PR adds quite a lot of opinionated set of new objects `Envelope` and fields. It's worth pointing out that introspecting the mailable can be done simply after calling `build()`, so this PR doesn't really "solve" much. It's an alternative syntax, which is more complex. I hope it works out.

**Comment by taylorotwell** [comment]
Added methods and conditionable trait to Envelope.

**Comment by Sammyjo20** [comment]
Methods look good!

**Comment by taylorotwell** [comment]
Documentation: https://github.com/laravel/docs/pull/8282

File: src/Illuminate/Mail/Mailable.php [hunk 6]
```diff
@@ -1424,20 +1426,10 @@ private function ensureEnvelopeIsHydrated()
...
-            $this->cc($cc->address, $cc->name);
-        }
-
-        foreach ($envelope->bcc as $bcc) {
-            $this->bcc($bcc->address, $bcc->name);
-        }
-
-        foreach ($envelope->replyTo as $replyTo) {
-            $this->replyTo($replyTo->address, $replyTo->name);
+        foreach (['to', 'cc', 'bcc', 'replyTo'] as $type) {
```
[the remark this answers is not in the document]
Comment (Kaleemullah007) [reply src/Illuminate/Mail/Mailable.php:1430]: New addition improves performance and avoid using nested loop.

**Comment by Asad-Ali-Haider** [comment]
How can we use this syntax into APIs?
